### PR TITLE
Fix for "CLI scripts print 'Application Instantiation Error' when redis cache enabled"

### PR DESCRIPTION
**Pull request by csthomas**
Pull Request for Issue #16903

### Summary of Changes
Fix for "CLI scripts print 'Application Instantiation Error' when redis cache enabled"

The error message comes from:
~~https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/factory.php#L122~~
https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Factory.php#L137

which means that application instance has not been created yet.

Because on redis handler I stop using `JFactory::getApplication()` then I had to change other related lines.

Because I can not use `$app` and I should not display error message to front end:
```php
if ($app->isClient('administrator'))
{
	JError::raiseWarning(500, 'Redis connection failed');
}
```

I changed it to:
```
throw new JCacheExceptionConnecting('Could not connect to redis server');
```

### Testing Instructions
Install joomla with cache Redis enabled.

Go to joomla folder and run `php -f cli/update_cron.php`

Before patch you get error message.
After patch there is no errors.

### Expected result
Script works without errors.

### Actual result
Get exception message: Application Instantiation Error

### Documentation Changes Required
I do not know.

### CODE REVIEW

File: libraries/joomla/cache/storage/redis.php [hunk 1]
```diff
@@ -55,73 +55,64 @@ public function __construct($options = array())
...
 	 *
 	 * @since   3.4
-	 * @note    As of 4.0 this method will throw a JCacheExceptionConnecting object on connection failure
 	 */
 	protected function getConnection()
 	{
 		if (static::isSupported() == false)
 		{
-			return false;
+			throw new RuntimeException('Redis Extension is not available');
```
Comment (laoneo) [inline libraries/joomla/cache/storage/redis.php]: Should you not throw here an `UnsupportedCacheException` exception?

Comment (csthomas) [reply libraries/joomla/cache/storage/redis.php]: fixed, thanks

**Comment by zero-24** [comment]
Please move the since tag to '__DEPLOY_VERSION__' ;)

**Comment by csthomas** [comment]
For a while I have a wrong commit with more changes, I removed that. Now IMO there is no place to put __DEPLOY_VERSION__

**Comment by zero-24** [comment]
Sorry looks like github fools me than a few minutes ago i saw a new method being added here now it is gone. Thanks for checking.

**Comment by csthomas** [comment]
There was a new method, but I removed it because it was not part of this PR.

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 59a36fd3e142f19fc9421e9ec770a5b16012547a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16904">issues.joomla.org/tracker/joomla-cms/16904</a>.</sub>

**Comment by csthomas** [comment]
I had to fix a conflict and I moved file changes to new location of redis file.

**Comment by csthomas** [comment]
Close this in favour #18395

**Pull request closed without merging**